content/writing-pipelines/index.md

```diff
@@ -115,9 +115,117 @@ files for [Vim](https://www.vim.org/), [Atom](https://atom.io/),
+
+Implicit conversion from `string` to `file` or `path` is permitted, as well as
+from `int` to `float` (though not the reverse).
+
+#### User-defined file types
+
+Users may define new file types using the `filetype` directive.  These behave
+like `file`, and can be implicitly converted to `string` or `file`, but not from
+one user file type to another.  This allows pipelines to be more clear about
+the format of files being passed around, and to exploit the type checking to
+ensure that files are being used conistently as they are passed around.
+
+#### Structured data types (Martian 4.0 preview)
+
+A `struct` is related to the same concepts in other languages like C, a named
+tuple in Python, or an object in javascript.  They can be declared,
+```
+struct MyType(
+    int   foo,
+    float bar,
+)
+```
+Members of a struct can be extracted using the familiar `.` syntax, e.g.
+```
+call FOO(
+    foo = STRUCT_OUTPUT.struct.foo,
+)
+```
+The output of a stage or pipeline is always a `struct`.  Because of this, the
+name of a stage or pipeline can be used as a type, to indicate a structure with
+the same members and types as the outputs of the stage or pipeline.
+
+Martian structures support a form of "[duck typing](https://en.wikipedia.org/wiki/Duck_typing)".
+If one has a struct `MyType` as declared above, and another type
+```
+struct MyBiggerType(
+    int    foo,
+    int    bar,
+    string baz,
+)
+```
+then a value of type `MyBiggerType` may be used for the input to a stage or
+pipeline which asks for a `MyType`.  This is because for every field in `MyType`
+there is a field with the same name in `MyBiggerType`, and that field in
+`MyBiggerType` has a type that is assignable to the type for that field in
+`MyType`.  Because of this, one can easily take a subset of the data from a
+`struct` with only the values one actually needs.  Values with struct types may
+also always be used for untyped map values.
+
+As an additional convenience, martian supports a "wildcard expansion" of a
+struct value when calling a stage, e.g.
+```
+call STAGE2(
+    foo = self.foo,
+    *   = STAGE1,
+)
+```
+This is equivalent to `input = STAGE1.output` for every output of `STAGE1` that
+is an input to `STAGE2`.  To prevent ambiguity, only one wildcard expansion is
+allowed for each call, and it is an error if one of the outputs of `STAGE1` was
+already assigned in the input call explicitly (e.g. `foo` in the example).
+
+#### Collection types
+
+Martian also supports collections of values as arrays or typed maps.  These are
+declared using a syntax that is familar to users of C-style languages.  Arrays
+are declared as for example `int[]`.  Typed maps (available in the martian 4.0
+preview) always have string keys, and
+are declared as for example `map<int>`.  These can be combined as for example
+`map<int[][]>[]`.
```

Comment:
Cool, I din't know this. Is 2d array the highest supported by martian?

Reply:
Array dimension is a signed 16-bit integer.  Please don't push the limits on that range :-)